### ISSUE TITLE
Visual Studio 2015 Build Issues

### ISSUE THREAD

**Issue by kiydev** [google/googletest#661]
I am wanting to use googlemock in my C++ project using Visual Studio 2015.  I have tried compiling version 1.7 and trunk, but keep running into issues.  Everything below is using trunk from Dec. 8th, 2015.

The first two problems seem rather simple:
1. gmock_config.props is using at an incorrect path for googletest

```
-      <GTestDir>../../gtest</GTestDir>
+      <GTestDir>../../../googletest</GTestDir>
```
1. gmock-internal-utils_test.cc has an error on including a header.  It is pointing to a header that is in googletest (not googlemock).  The following is the change I made to continue

```
-  #include "src/gtest-internal-inl.h"
+  #include "../src/gtest-internal-inl.h"
```

However, I am stuck on an issue that someone else has had (link to same issue from [google group](https://groups.google.com/forum/#!topic/googlemock/SSUE5ywM3R4)).  Here is the output from MSVC:

```
1>------ Build started: Project: gmock_test, Configuration: Debug Win32 ------
1>  gmock_all_test.cc
1> googletest-master\googlemock\include\gmock/gmock-matchers.h(555): error C2440: 'return': cannot convert from 'const testing::internal::EqMatcher<testing::gmock_matchers_test::ConvertibleFromAny>' to 'testing::Matcher<testing::gmock_matchers_test::ConvertibleFromAny>'
1> googletest-master\googlemock\include\gmock/gmock-matchers.h(555): note: No constructor could take the source type, or constructor overload resolution was ambiguous
1> googletest-master\googlemock\include\gmock/gmock-matchers.h(546): note: while compiling class template member function 'testing::Matcher<testing::gmock_matchers_test::ConvertibleFromAny> testing::internal::MatcherCastImpl<T,M>::CastImpl(const M &,testing::internal::BooleanConstant<true>)'
1>          with
1>          [
1>              T=testing::gmock_matchers_test::ConvertibleFromAny,
1>              M=testing::internal::EqMatcher<testing::gmock_matchers_test::ConvertibleFromAny>
1>          ]
1>  googletest-master\googlemock\include\gmock/gmock-matchers.h(534): note: see reference to function template instantiation 'testing::Matcher<testing::gmock_matchers_test::ConvertibleFromAny> testing::internal::MatcherCastImpl<T,M>::CastImpl(const M &,testing::internal::BooleanConstant<true>)' being compiled
1>          with
1>          [
1>              T=testing::gmock_matchers_test::ConvertibleFromAny,
1>              M=testing::internal::EqMatcher<testing::gmock_matchers_test::ConvertibleFromAny>
1>          ]
1>  googletest-master\googlemock\include\gmock/gmock-matchers.h(611): note: see reference to class template instantiation 'testing::internal::MatcherCastImpl<T,M>' being compiled
1>          with
1>          [
1>              T=testing::gmock_matchers_test::ConvertibleFromAny,
1>              M=testing::internal::EqMatcher<testing::gmock_matchers_test::ConvertibleFromAny>
1>          ]
1>  googletest-master\googlemock\test/gmock-matchers_test.cc(638): note: see reference to function template instantiation 'testing::Matcher<testing::gmock_matchers_test::ConvertibleFromAny> testing::MatcherCast<testing::gmock_matchers_test::ConvertibleFromAny,testing::internal::EqMatcher<testing::gmock_matchers_test::ConvertibleFromAny>>(const M &)' being compiled
1>          with
1>          [
1>              M=testing::internal::EqMatcher<testing::gmock_matchers_test::ConvertibleFromAny>
1>          ]
========== Build: 0 succeeded, 1 failed, 2 up-to-date, 0 skipped ==========
```

The problem is with gmock_matchers.h:

```
  static Matcher<T> CastImpl(const M& polymorphic_matcher_or_value,
                             BooleanConstant<true>) {
    // M is implicitly convertible to Matcher<T>, which means that either
    // M is a polymorhpic matcher or Matcher<T> has an implicit constructor
    // from M.  In both cases using the implicit conversion will produce a
    // matcher.
    //
    // Even if T has an implicit constructor from M, it won't be called because
    // creating Matcher<T> would require a chain of two user-defined conversions
    // (first to create T from M and then to create Matcher<T> from T).
    return polymorphic_matcher_or_value;
  }
```

A workaround  for this would be greatly appreciated.  Any ideas?

**Comment by wtjones** [comment]
I am having some success with cmake via this post: https://groups.google.com/d/msg/googlemock/SSUE5ywM3R4/6wBHa11kDgAJ

> run "cmake -Dgtest_build_samples=ON" from inside the Google Mock directory

Use the sln that it creates. Everything builds in x86 although when I switched it to x64 I had to manually just build the main libs.

I am also not sure why it outputs `gtest_main.lib` instead of `gtest_maind.lib` (the one my project uses) but I am uncertain if this matters yet.

**Comment by kiydev** [comment]
Thanks for the suggestion.

With CMake (which I would prefer to not to have to use) I was able to build the solution files for Visual Studio 2015.

By default it builds static libraries.  You can create dlls by specifying
```
-DBUILD_SHARED_LIBS=ON
```

A gotcha for me was that the static libraries were also statically linking the C++ runtime, which was causing problems with the Visual Studio test process (was getting some strange assertion failures).  I suggest changing to dynamically linking the C++ runtime (see [MSDN](https://msdn.microsoft.com/en-us/library/2kzt1wy3.aspx) for changing this for each project).

I had to manually create x64 configurations (maybe there was a flag for CMake do this?).  The most annoying part of this was that CMake put the x86 machine target in the additional options, so, when I created the x64 configurations it was still targeting x86 (I had to manually remove that option from each project's additional options to the compiler).

After all this, though, I have google mock in my project with Visual Studio 2015 using the Visual Studio test framework... and it's working!

**Comment by wtjones** [comment]
I am using the .lib files, and I noticed that debugging a test will crash the VS test runtime at the end after the asserts. I can debug up that point just fine, and running a test without debug works fine. This started after moving to VS 2015 and rebuilding gmock.

I may try using dlls instead as you mentioned.

**Comment by paulsapps** [comment]
So what was the actual fix for the build error?

**Comment by wtjones** [comment]
Using cmake to regen was the fix for the build error. Give that a shot.

 I remember additional fiddling, but the specifics escape me at the moment.

**Comment by aregier** [comment]
We (Lexmark Enterprise Software) would also appreciate a fix for this issue so that the static libraries could be compiled and used from Visual Studio 2015.

**Comment by kiydev** [comment]
It is possible to compile the static (and the dynamic as well) libraries
using cmake. However, I ran into issues using the static libraries due to
them statically linking the C++ runtime (see my comment above). I believe
(though I don't quite remember what I actually ended up doing) once you
build the static libraries with dynamic linking of the C++ runtime, VS test
framework will work just great.

**Comment by KindDragon** [comment]
I think it should be fixed in PR #723

**Comment by aaicken** [comment]
I've just tried building googlemock in Visual C++ Express 2010 using the latest version 1.8.0 package and was getting this error:

`gmock-internal-utils_test.cc(54): fatal error C1083: Cannot open include file: 'src/gtest-internal-inl.h': No such file or directory`

I fixed it by modifying the gmock_config.props file:

```
-       <AdditionalIncludeDirectories>$(GTestDir)/include;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
+       <AdditionalIncludeDirectories>$(GTestDir)/include;$(GTestDir);%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
```

**Comment by KindDragon** [comment]
> I've just tried building googlemock in Visual C++ Express 2010 using the latest version 1.8.0 package and was getting this error:

You can use CMake instead, but it should be fixed in PR #852
